File: server/app/services/bullet_service.py
"""Bullet service — CRUD operations for bullets."""

import time

import aiosqlite

from app.models.bullet import BulletResponse
from app.services.tag_service import sync_tags


def _row_to_bullet(row: aiosqlite.Row) -> BulletResponse:
    return BulletResponse(
        id=row["id"],
        document_id=row["document_id"],
        parent_id=row["parent_id"],
        content=row["content"],
        position=row["position"],
        is_complete=bool(row["is_complete"]),
        created_at=row["created_at"],
        updated_at=row["updated_at"],
    )


async def get_document_bullets(
    db: aiosqlite.Connection,
    document_id: str,
) -> list[BulletResponse] | None:
    """
    Return a flat list of all non-deleted bullets for a document, ordered by position.
    Returns None if the document does not exist.
    """
    doc_rows = await db.execute_fetchall(
        "SELECT id FROM documents WHERE id = ? AND deleted_at IS NULL",
        (document_id,),
    )
    if not doc_rows:
        return None

    rows = await db.execute_fetchall(
        "SELECT id, document_id, parent_id, content, position, is_complete, "
        "created_at, updated_at "
        "FROM bullets "
        "WHERE document_id = ? AND deleted_at IS NULL "
        "ORDER BY position",
        (document_id,),
    )
    return [_row_to_bullet(r) for r in rows]


async def create_bullet(
    db: aiosqlite.Connection,
    bullet_id: str,
    document_id: str,
    parent_id: str | None,
    content: str,
    position: str,
    is_complete: bool,
) -> BulletResponse:
    """
    Create a new bullet.  The bullet_id is client-supplied (UUID v4) to
    support idempotency.  Uses INSERT OR IGNORE so a duplicate call is
    silently a no-op, and then fetches the existing row.
    """
    now = int(time.time() * 1000)

    await db.execute(
        "INSERT OR IGNORE INTO bullets "
        "(id, document_id, parent_id, content, position, is_complete, created_at, updated_at) "
        "VALUES (?, ?, ?, ?, ?, ?, ?, ?)",
        (bullet_id, document_id, parent_id, content, position, int(is_complete), now, now),
    )
    await sync_tags(db, bullet_id, content)
    await db.commit()

    rows = await db.execute_fetchall(
        "SELECT id, document_id, parent_id, content, position, is_complete, "
        "created_at, updated_at FROM bullets WHERE id = ?",
        (bullet_id,),
    )
    return _row_to_bullet(rows[0])


async def update_bullet(
    db: aiosqlite.Connection,
    bullet_id: str,
    content: str | None,
    position: str | None,
    parent_id: str | None,  # None means "don't change"; use a sentinel for explicit null
    is_complete: bool | None,
    clear_parent: bool = False,
) -> BulletResponse | None:
    """
    Update one or more fields of a bullet.

    `clear_parent=True` sets parent_id to NULL (move to root level).
    Returns None if bullet not found.
    """
    rows = await db.execute_fetchall(
        "SELECT id, document_id, parent_id, content, position, is_complete, "
        "created_at, updated_at FROM bullets WHERE id = ? AND deleted_at IS NULL",
        (bullet_id,),
    )
    if not rows:
        return None

    current = rows[0]
    new_content = content if content is not None else current["content"]
    new_position = position if position is not None else current["position"]
    new_is_complete = int(is_complete) if is_complete is not None else current["is_complete"]
    now = int(time.time() * 1000)

    if clear_parent:
        new_parent_id = None
    elif parent_id is not None:
        new_parent_id = parent_id
    else:
        new_parent_id = current["parent_id"]

    await db.execute(
        "UPDATE bullets SET content = ?, position = ?, parent_id = ?, "
        "is_complete = ?, updated_at = ? WHERE id = ?",
        (new_content, new_position, new_parent_id, new_is_complete, now, bullet_id),
    )
    await sync_tags(db, bullet_id, new_content)
    await db.commit()

    updated = await db.execute_fetchall(
        "SELECT id, document_id, parent_id, content, position, is_complete, "
        "created_at, updated_at FROM bullets WHERE id = ?",
        (bullet_id,),
    )
    return _row_to_bullet(updated[0])


async def delete_bullet(db: aiosqlite.Connection, bullet_id: str) -> bool:
    """
    Soft-delete a bullet and cascade to all descendants by recursively
    soft-deleting children.  Returns False if the bullet was not found.
    """
    rows = await db.execute_fetchall(
        "SELECT id FROM bullets WHERE id = ? AND deleted_at IS NULL",
        (bullet_id,),
    )
    if not rows:
        return False

    now = int(time.time() * 1000)
    await _soft_delete_recursive(db, bullet_id, now)
    await db.commit()
    return True


async def _soft_delete_recursive(
    db: aiosqlite.Connection, bullet_id: str, now: int
) -> None:
    """Recursively soft-delete a bullet and all its non-deleted descendants."""
    # Soft-delete the bullet itself.
    await db.execute(
        "UPDATE bullets SET deleted_at = ? WHERE id = ? AND deleted_at IS NULL",
        (now, bullet_id),
    )

    # Find direct children and recurse.
    children = await db.execute_fetchall(
        "SELECT id FROM bullets WHERE parent_id = ? AND deleted_at IS NULL",
        (bullet_id,),
    )
    for child in children:
        await _soft_delete_recursive(db, child["id"], now)

File: server/src/services/bulletService.ts
import { db as defaultDb } from '../../db/index.js';
import { bullets } from '../../db/schema.js';
import { eq, and, isNull, asc } from 'drizzle-orm';
import { randomUUID } from 'node:crypto';
import { recordUndoEvent } from './undoService.js';
import type { DB } from '../../db/index.js';
import type { BulletRow, UndoOp } from './undoService.js';

// Full bullet shape as returned from DB
export type Bullet = {
  id: string;
  documentId: string;
  userId: string;
  parentId: string | null;
  content: string;
  position: number;
  isComplete: boolean;
  isCollapsed: boolean;
  note: string | null;
  deletedAt: Date | null;
  createdAt: Date;
  updatedAt: Date;
};

// ─── Helpers ────────────────────────────────────────────────────────────────

/** Load a bullet by id+userId, optionally requiring it to be non-deleted. */
async function loadBullet(
  dbInstance: DB,
  bulletId: string,
  userId: string,
  opts: { requireActive?: boolean } = {}
): Promise<Bullet | null> {
  const conditions = opts.requireActive
    ? and(eq(bullets.id, bulletId), eq(bullets.userId, userId), isNull(bullets.deletedAt))
    : and(eq(bullets.id, bulletId), eq(bullets.userId, userId));
  const row = await dbInstance.query.bullets.findFirst({ where: conditions });
  return (row as Bullet) ?? null;
}

/** Update a bullet inside a transaction and record an undo event. */
async function updateWithUndo(
  dbInstance: DB,
  userId: string,
  bulletId: string,
  set: Record<string, unknown>,
  actionName: string,
  forwardOp: UndoOp,
  inverseOp: UndoOp
): Promise<Bullet> {
  let updated: Bullet | undefined;
  await dbInstance.transaction(async (tx) => {
    const rows = await tx
      .update(bullets)
      .set(set as Parameters<ReturnType<typeof tx.update>['set']>[0])
      .where(eq(bullets.id, bulletId))
      .returning();
    updated = rows[0] as Bullet;
    await recordUndoEvent(tx, userId, actionName, forwardOp, inverseOp);
  });
  return updated!;
}

// ─── Public API ─────────────────────────────────────────────────────────────

/**
 * Compute FLOAT8 midpoint insert position for a bullet among siblings.
 */
export async function computeBulletInsertPosition(
  dbInstance: DB,
  documentId: string,
  parentId: string | null,
  afterId: string | null
): Promise<number> {
  const siblings = await dbInstance
    .select({ id: bullets.id, position: bullets.position })
    .from(bullets)
    .where(
      and(
        eq(bullets.documentId, documentId),
        parentId != null ? eq(bullets.parentId, parentId) : isNull(bullets.parentId),
        isNull(bullets.deletedAt)
      )
    )
    .orderBy(asc(bullets.position));

  if (siblings.length === 0) return 1.0;

  if (afterId === null) {
    return siblings[0].position / 2;
  }

  const afterIdx = siblings.findIndex(s => s.id === afterId);
  if (afterIdx === -1) {
    return siblings[siblings.length - 1].position + 1.0;
  }

  const prev = siblings[afterIdx].position;
  const next = siblings[afterIdx + 1]?.position;
  return next !== undefined ? (prev + next) / 2 : prev + 1.0;
}

/**
 * Return all non-deleted bullets for a document, ordered by position.
 */
export async function getDocumentBullets(
  userId: string,
  documentId: string,
  dbInstance: DB = defaultDb
): Promise<Bullet[]> {
  return dbInstance
    .select()
    .from(bullets)
    .where(
      and(
        eq(bullets.documentId, documentId),
        eq(bullets.userId, userId),
        isNull(bullets.deletedAt)
      )
    )
    .orderBy(asc(bullets.position)) as Promise<Bullet[]>;
}

/**
 * Create a new bullet at the computed position, recording an undo event.
 */
export async function createBullet(
  userId: string,
  {
    documentId,
    parentId,
    afterId,
    content = '',
  }: {
    documentId: string;
    parentId?: string | null;
    afterId?: string | null;
    content?: string;
  },
  dbInstance: DB = defaultDb
): Promise<Bullet> {
  const resolvedParentId = parentId ?? null;
  const resolvedAfterId = afterId ?? null;

  const position = await computeBulletInsertPosition(
    dbInstance,
    documentId,
    resolvedParentId,
    resolvedAfterId
  );

  const id = randomUUID();
  let created: Bullet | undefined;

  await dbInstance.transaction(async (tx) => {
    const inserted = await tx
      .insert(bullets)
      .values({
        id,
        userId,
        documentId,
        parentId: resolvedParentId,
        content,
        position,
        isComplete: false,
        isCollapsed: false,
      })
      .returning();

    created = inserted[0] as Bullet;

    await recordUndoEvent(
      tx,
      userId,
      'create_bullet',
      { type: 'restore_bullet', id },
      { type: 'restore_bullet_delete', id }
    );
  });

  return created!;
}

/**
 * Indent a bullet: makes it the last child of its previous sibling.
 * No-op if the bullet is already the first sibling.
 */
export async function indentBullet(
  userId: string,
  bulletId: string,
  dbInstance: DB = defaultDb
): Promise<Bullet | null> {
  const bullet = await loadBullet(dbInstance, bulletId, userId, { requireActive: true });
  if (!bullet) return null;

  // Get siblings to find the previous sibling
  const siblings = await dbInstance
    .select({ id: bullets.id, position: bullets.position })
    .from(bullets)
    .where(
      and(
        eq(bullets.documentId, bullet.documentId),
        bullet.parentId != null
          ? eq(bullets.parentId, bullet.parentId)
          : isNull(bullets.parentId),
        isNull(bullets.deletedAt)
      )
    )
    .orderBy(asc(bullets.position));

  const index = siblings.findIndex(s => s.id === bulletId);
  if (index === 0) return bullet; // Already first sibling — no-op

  const previousSibling = siblings[index - 1];

  // Find last child of previousSibling to append after it
  const prevSiblingChildren = await dbInstance
    .select({ id: bullets.id, position: bullets.position })
    .from(bullets)
    .where(
      and(
        eq(bullets.documentId, bullet.documentId),
        eq(bullets.parentId, previousSibling.id),
        isNull(bullets.deletedAt)
      )
    )
    .orderBy(asc(bullets.position));

  const lastChildId =
    prevSiblingChildren.length > 0
      ? prevSiblingChildren[prevSiblingChildren.length - 1].id
      : null;

  const appendPosition = await computeBulletInsertPosition(
    dbInstance,
    bullet.documentId,
    previousSibling.id,
    lastChildId
  );

  return updateWithUndo(
    dbInstance, userId, bulletId,
    { parentId: previousSibling.id, position: appendPosition },
    'indent_bullet',
    { type: 'move_bullet', id: bulletId, parentId: previousSibling.id, position: appendPosition },
    { type: 'move_bullet', id: bulletId, parentId: bullet.parentId, position: bullet.position }
  );
}

/**
 * Outdent a bullet: makes it a sibling of its parent, positioned after parent.
 * No-op if the bullet is already at root level.
 */
export async function outdentBullet(
  userId: string,
  bulletId: string,
  dbInstance: DB = defaultDb
): Promise<Bullet | null> {
  const bullet = await loadBullet(dbInstance, bulletId, userId, { requireActive: true });
  if (!bullet) return null;
  if (!bullet.parentId) return bullet; // Already root level — no-op

  const parent = await dbInstance.query.bullets.findFirst({
    where: eq(bullets.id, bullet.parentId),
  });
  if (!parent) return bullet;

  const newPosition = await computeBulletInsertPosition(
    dbInstance,
    bullet.documentId,
    parent.parentId,
    parent.id
  );

  return updateWithUndo(
    dbInstance, userId, bulletId,
    { parentId: parent.parentId, position: newPosition },
    'outdent_bullet',
    { type: 'move_bullet', id: bulletId, parentId: parent.parentId, position: newPosition },
    { type: 'move_bullet', id: bulletId, parentId: bullet.parentId, position: bullet.position }
  );
}

/**
 * Load all descendants of a bullet (recursive via parentId chain).
 * Uses a Map for O(n) traversal instead of repeated filtering.
 */
async function getDescendantIds(
  dbInstance: DB,
  bulletId: string,
  documentId: string
): Promise<Set<string>> {
  const all = await dbInstance
    .select({ id: bullets.id, parentId: bullets.parentId })
    .from(bullets)
    .where(and(eq(bullets.documentId, documentId), isNull(bullets.deletedAt)));

  // Build parent→children index for O(n) walk
  const childrenOf = new Map<string | null, string[]>();
  for (const b of all) {
    const key = b.parentId;
    let arr = childrenOf.get(key);
    if (!arr) { arr = []; childrenOf.set(key, arr); }
    arr.push(b.id);
  }

  const result = new Set<string>();
  const queue = [bulletId];
  while (queue.length > 0) {
    const current = queue.shift()!;
    const children = childrenOf.get(current) ?? [];
    for (const childId of children) {
      result.add(childId);
      queue.push(childId);
    }
  }

  return result;
}

/**
 * Move a bullet and all its descendants to a new parent/position.
 * Throws if the move would create a cycle (moving under own descendant).
 */
export async function moveBullet(
  userId: string,
  bulletId: string,
  { newParentId, afterId }: { newParentId: string | null; afterId: string | null },
  dbInstance: DB = defaultDb
): Promise<Bullet | null> {
  const bullet = await loadBullet(dbInstance, bulletId, userId, { requireActive: true });
  if (!bullet) return null;

  if (newParentId !== null) {
    const descendants = await getDescendantIds(dbInstance, bulletId, bullet.documentId);
    if (descendants.has(newParentId)) {
      throw new Error('Cannot move a bullet under one of its own descendants');
    }
  }

  const newPosition = await computeBulletInsertPosition(
    dbInstance,
    bullet.documentId,
    newParentId,
    afterId
  );

  return updateWithUndo(
    dbInstance, userId, bulletId,
    { parentId: newParentId, position: newPosition },
    'move_bullet',
    { type: 'move_bullet', id: bulletId, parentId: newParentId, position: newPosition },
    { type: 'move_bullet', id: bulletId, parentId: bullet.parentId, position: bullet.position }
  );
}

/**
 * Soft-delete a bullet by setting deletedAt = now().
 */
export async function softDeleteBullet(
  userId: string,
  bulletId: string,
  dbInstance: DB = defaultDb
): Promise<Bullet | null> {
  const bullet = await loadBullet(dbInstance, bulletId, userId, { requireActive: true });
  if (!bullet) return null;

  return updateWithUndo(
    dbInstance, userId, bulletId,
    { deletedAt: new Date() },
    'delete_bullet',
    { type: 'delete_bullet', id: bulletId },
    { type: 'restore_bullet', id: bulletId }
  );
}

/**
 * Mark or unmark a bullet complete.
 */
export async function markComplete(
  userId: string,
  bulletId: string,
  isComplete: boolean,
  dbInstance: DB = defaultDb
): Promise<Bullet | null> {
  const bullet = await loadBullet(dbInstance, bulletId, userId);
  if (!bullet) return null;

  return updateWithUndo(
    dbInstance, userId, bulletId,
    { isComplete },
    'mark_complete',
    { type: 'update_bullet', id: bulletId, fields: { isComplete } },
    { type: 'update_bullet', id: bulletId, fields: { isComplete: bullet.isComplete } }
  );
}

/**
 * Collapse or expand a bullet.
 */
export async function setCollapsed(
  userId: string,
  bulletId: string,
  isCollapsed: boolean,
  dbInstance: DB = defaultDb
): Promise<Bullet | null> {
  const bullet = await loadBullet(dbInstance, bulletId, userId);
  if (!bullet) return null;

  return updateWithUndo(
    dbInstance, userId, bulletId,
    { isCollapsed },
    'set_collapsed',
    { type: 'update_bullet', id: bulletId, fields: { isCollapsed } },
    { type: 'update_bullet', id: bulletId, fields: { isCollapsed: bullet.isCollapsed } }
  );
}

/**
 * Patch a bullet's note field.
 */
export async function patchBullet(
  userId: string,
  bulletId: string,
  fields: { note?: string | null },
  dbInstance: DB = defaultDb
): Promise<Bullet | null> {
  const bullet = await loadBullet(dbInstance, bulletId, userId);
  if (!bullet) return null;

  const note = fields.note !== undefined ? (fields.note ?? null) : undefined;
  if (note === undefined) return bullet; // nothing to update

  return updateWithUndo(
    dbInstance, userId, bulletId,
    { note, updatedAt: new Date() },
    'update_note',
    { type: 'update_bullet', id: bulletId, fields: { note } as unknown as Partial<BulletRow> },
    { type: 'update_bullet', id: bulletId, fields: { note: bullet.note ?? null } as unknown as Partial<BulletRow> }
  );
}
